include/lbann/layers/loss/cross_entropy.hpp

```diff
@@ -0,0 +1,139 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2016, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_LAYERS_LOSS_CROSS_ENTROPY_HPP_INCLUDED
+#define LBANN_LAYERS_LOSS_CROSS_ENTROPY_HPP_INCLUDED
+
+#include "lbann/layers/layer.hpp"
+
+namespace lbann {
+
+template <data_layout T_layout, El::Device Dev>
+class cross_entropy_layer : public Layer {
+public:
+
+  cross_entropy_layer(lbann_comm *comm) : Layer(comm) {
+    // Expects inputs for prediction and ground truth
+    m_expected_num_parent_layers = 2;
+  }
+
+  cross_entropy_layer* copy() const override { return new cross_entropy_layer(*this); }
+  std::string get_type() const override { return "cross entropy"; }
+  data_layout get_data_layout() const override { return T_layout; }
+  El::Device get_device_allocation() const override { return Dev; }
+
+  void setup_dims() override {
+    Layer::setup_dims();
+    this->m_num_neurons = 1;
+    this->m_num_neuron_dims = 1;
+    this->m_neuron_dims = {1};
+  }
+
+  void fp_compute() override {
+
+    // Initialize workspace
+    std::unique_ptr<AbsDistMat> workspace;
+    const auto& prediction = get_prev_activations(0);
+    switch (get_data_layout()) {
+    case data_layout::DATA_PARALLEL:
+      workspace.reset(new StarVCMat<Dev>(prediction.Grid(),
+                                         prediction.Root()));
+      break;
+    case data_layout::MODEL_PARALLEL:
+      workspace.reset(new StarMRMat<Dev>(prediction.Grid(),
+                                         prediction.Root()));
+      break;
+    default: LBANN_ERROR("invalid data layout");
+    }
```

Comment:
Why do we allocate a distributed matrix every time we do forward prop? That seems expensive.

Reply:
It's not that bad since we only allocate memory equal to the local mini-batch size. It's less of a problem on GPUs since we use CUB's memory pool. Since the layer deallocates the memory at the end of its computation, it's also (very very barely) more efficient with GPU memory.

Reply:
Understood that it is not that bad, but why aren't these matrices just setup in the setup_data function?